### PR TITLE
Wip/hughsie/save events

### PR DESCRIPTION
**Pull request by hughsie**
Type of pull request:

- [ ] New plugin (Please include [new plugin checklist](https://github.com/fwupd/fwupd/wiki/New-plugin-checklist))
- [X] Code fix
- [ ] Feature
- [ ] Documentation

### CODE REVIEW

**Comment by hughsie** [comment]
@fdanis-oss

File: src/fu-usb-backend.c [hunk 1]
```diff
@@ -231,11 +235,22 @@ fu_usb_backend_init(FuUsbBackend *self)
 {
 }
 
+static void
+fu_colorhug_plugin_constructed(GObject *obj)
```
Comment (fdanis-oss) [inline src/fu-usb-backend.c:239]: Shouldn't it be:
```suggestion
fu_usb_backend_constructed(GObject *obj)
```

File: src/fu-usb-backend.c [hunk 2]
```diff
@@ -231,11 +235,22 @@ fu_usb_backend_init(FuUsbBackend *self)
...
+			 G_CALLBACK(fu_usb_backend_context_flags_notify_cb),
+			 self);
+}
+
 static void
 fu_usb_backend_class_init(FuUsbBackendClass *klass)
 {
 	GObjectClass *object_class = G_OBJECT_CLASS(klass);
 	FuBackendClass *klass_backend = FU_BACKEND_CLASS(klass);
+	object_class->constructed = fu_colorhug_plugin_constructed;
```
Comment (fdanis-oss) [inline src/fu-usb-backend.c:253]: see comment above

File: src/fu-usb-backend.c (hunk 1 above)
Comment (hughsie) [reply src/fu-usb-backend.c]: Ooops! Thanks, fixed.

**Comment by fdanis-oss** [comment]
This currently fails when running `fwupdtool --save-backends=/tmp/test.json install e1b2e0294ffa8ff056bea9669d839d17f40ae0545809664f654b03f932b90bc7-CalDigit-TS4-combind-121022-secure-Cus-37.cab`:
```
0x00007ffff75e1879 in g_usb_context_set_flags (self=0x0, flags=G_USB_CONTEXT_FLAGS_SAVE_EVENTS) at ../subprojects/gusb/gusb/gusb-context.c:800
800		priv->flags = flags;
```

**Comment by hughsie** [comment]
> This currently fails

Try this new version pls?

**Comment by fdanis-oss** [comment]
no more crashes but usb events are no more saved:
```
Thread 1 "fwupdtool" hit Breakpoint 1, fu_usb_backend_context_flags_check (self=0x7ffff7dca9ef) at ../src/fu-usb-backend.c:121
…
126		if (self->usb_ctx != NULL && fu_context_has_flag(ctx, FU_CONTEXT_FLAG_SAVE_EVENTS)) {
(gdb) print *self
$1 = {parent_instance = {parent_instance = {g_type_instance = {Python Exception <class 'gdb.error'> No type named TypeNode.: 
g_class = }, ref_count = 1, qdata = 0x5555555d2f00}}, usb_ctx = 0x0}
(gdb) bt
#0  fu_usb_backend_context_flags_check (self=0x5555555c0660) at ../src/fu-usb-backend.c:126
#1  0x00007ffff7dcb2de in fu_usb_backend_constructed (obj=0x5555555c0660) at ../src/fu-usb-backend.c:250
#2  0x00007ffff79a3526 in  () at /usr/lib/x86_64-linux-gnu/libgobject-2.0.so.0
#3  0x00007ffff79a5378 in g_object_new_valist () at /usr/lib/x86_64-linux-gnu/libgobject-2.0.so.0
#4  0x00007ffff79a56cd in g_object_new () at /usr/lib/x86_64-linux-gnu/libgobject-2.0.so.0
#5  0x00007ffff7dcb412 in fu_usb_backend_new (ctx=0x7fffe400e6d0) at ../src/fu-usb-backend.c:274
#6  0x00007ffff7db97a7 in fu_engine_init (self=0x5555555d30c0) at ../src/fu-engine.c:8451
#7  0x00007ffff79c41dd in g_type_create_instance () at /usr/lib/x86_64-linux-gnu/libgobject-2.0.so.0
#8  0x00007ffff79a334d in  () at /usr/lib/x86_64-linux-gnu/libgobject-2.0.so.0
#9  0x00007ffff79a4b45 in g_object_new_with_properties () at /usr/lib/x86_64-linux-gnu/libgobject-2.0.so.0
#10 0x00007ffff79a56f1 in g_object_new () at /usr/lib/x86_64-linux-gnu/libgobject-2.0.so.0
#11 0x00007ffff7db9e50 in fu_engine_new () at ../src/fu-engine.c:8573
#12 0x00005555555697f5 in main (argc=3, argv=0x7fffffffe388) at ../src/fu-tool.c:4082
```

**Comment by hughsie** [comment]
Right, I think you'll need to set FU_CONTEXT_FLAG_SAVE_EVENTS unconditionally in your branch. I have other ideas longer term.

**Comment by fdanis-oss** [comment]
Is there anything preventing to call `fu_usb_backend_context_flags_check()` at the end of `fu_usb_backend_setup()`?

**Comment by hughsie** [comment]
> at the end of fu_usb_backend_setup

Not sure I understand the question; by _setup() we should be watching the context flags and it should notify.

**Review by fdanis-oss — APPROVED** [review]

**Pull request merged**